### PR TITLE
Fix breakpoints on execute trapped page

### PR DESCRIPTION
**Pull request by masthoon**
Fix for Issue #192

### CODE REVIEW

**Comment by tklengyel** [comment]
So this looks fine, however I'm thinking that it may still have a hidden issue. An instruction fetch may load up to 8-bytes from memory, however, we only check if the address of the first byte is one we trapped. An int3 instruction may have arbitrary number of prefixes that don't actually change the behavior of the instruction, so it may be possible that our int3 is injected after a prefix, but the memory execute event will be at the address of the prefix thus skipping the int3 simulation here. Maybe we should just go from s->memaccess.pa up to s->memaccess.pa + 7 and check each of those addresses if they are trapped locations with int3.

**Comment by tklengyel** [comment]
This is ok to test.

File: src/libdrakvuf/vmi.c [hunk 1]
```diff
@@ -300,6 +300,29 @@ event_response_t pre_mem_cb(vmi_instance_t vmi, vmi_event_t *event) {
 
         loop = loop->next;
     }
+
+    /* We need to call breakpoint handlers registered for this physical address */
+    struct wrapper *sbp = g_hash_table_lookup(drakvuf->breakpoint_lookup_pa, &s->memaccess.pa);
+    if (event->mem_event.out_access & VMI_MEMACCESS_X && sbp) {
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c]: Probably better to do the hashtable lookup only after the check for VMI_MEMACCESS_X for performance reasons.

File: src/libdrakvuf/vmi.c [hunk 2]
```diff
@@ -300,6 +300,29 @@ event_response_t pre_mem_cb(vmi_instance_t vmi, vmi_event_t *event) {
...
+        reg_t cr3 = event->regs.x86->cr3;
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c]: No need to have this assigned to a local variable.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by masthoon** [comment]
I made the changes according to your comments.
In my opinion, breakpoints aren't supposed to be placed at the middle of an instruction or in self-modifying code, so this shouldn't be an issue. All of the plugins use breakpoints on function entry and if you misplace a breakpoint, the guest program will probably crash.
In which cases do you think this can happen ? 
By the way, the code handling breakpoints in _int3_cb_ is similar and it is affected by the same issue.

**Comment by tklengyel** [comment]
Thanks!

We indeed tend to inject breakpoints into function entry points but that's not a hard requirement. Yes, it's a pain to figure out where else to inject it - like injecting it into the middle of an another instruction is not going to go well but you could also just use a disassembler to help you figure out where it is safe to inject.

And yes, the int3_cb has a similar issue when we get an int3 that wasn't injected by us. If it's an int3 with a prefix, we are in trouble as we swallow the int3 (Ignoring old breakpoint event found in the queue).

**Comment by tklengyel** [comment]
Retest this please.

**Pull request merged**